src/RemediationRoleSaver.cpp

```diff
@@ -79,35 +151,194 @@ void RemediationSaverBase::saveToFile(const QString& filename)
+    ProfileBasedRemediationSaver(parentWindow, session,
+            puppetSaveMessage, puppetFiletypeExtension, puppetFiletypeTemplate, puppetFixType)
+{}
+
+
+#ifndef SCAP_WORKBENCH_USE_LIBRARY_FOR_RESULT_BASED_REMEDIATION_ROLES_GENERATION
+ResultBasedProcessRemediationSaver::ResultBasedProcessRemediationSaver(QWidget* parentWindow, const QByteArray& arfContents,
+        const QString& saveMessage, const QString& filetypeExtension, const QString& filetypeTemplate, const QString& fixType):
+    RemediationSaverBase(parentWindow, saveMessage, filetypeExtension, filetypeTemplate, fixType)
 {
-    // TODO: Add guess that uses benchmark and profile names
-    return QString("remediation");
+    mArfFile.setAutoRemove(true);
+    mArfFile.open();
+    mArfFile.write(arfContents);
+    mArfFile.close();
 }
 
 
-BashRemediationSaver::BashRemediationSaver(QWidget* parentWindow, ScanningSession* session):RemediationSaverBase(parentWindow, session)
+void ResultBasedProcessRemediationSaver::saveToFile(const QString& filename)
 {
-    mSaveMessage = QObject::tr("Save remediation role as a bash script");
-    mFiletypeExtension = "sh";
-    mFiletypeTemplate = QObject::tr("bash script (*.%1)");
-    mFixTemplate = QObject::tr("urn:xccdf:fix:script:sh");
+    QStringList args;
+    args.append("xccdf");
+    args.append("generate");
+    args.append("fix");
+
+    args.append("--template");
+    args.append(mTemplateString);
+    args.append("--output");
+    args.append(filename);
+
+    // vvv This will work, if there is only one result ID in the ARF file, it will be picked no matter what the argument value is.
+    // However, ommitting --result-id "" won't work.
+    args.append("--result-id");
+    args.append("");
+
+    args.append(mArfFile.fileName());
+
+    // Launching a process and going through its output is something we do already in OscapScannerLocal::evaluate()
+    // This is a lightweight launch though.
+    QProcess process(mParentWindow);
+
+    TemporaryDir workingDir;
+    process.setWorkingDirectory(workingDir.getPath());
+    QString program(SCAP_WORKBENCH_LOCAL_OSCAP_PATH);
+
+    process.start(program, args);
+    process.waitForStarted();
+
+    const unsigned int remediationGenerationTimeout = 10000;
+
+    const int process_finished_on_time = process.waitForFinished(remediationGenerationTimeout);
+
+    if (!process_finished_on_time)
+    {
+        QString message = QObject::tr("The process that was supposed to generate remediations didn't finish on time (i.e. within %1 secs), so it was terminated.").arg(remediationGenerationTimeout / 1000);
+        process.kill();
+        throw std::runtime_error(message.toUtf8().constData());
+    }
+
+    if (process.exitCode() != 0)
+    {
+        throw std::runtime_error(QObject::tr("There was an error in course of remediation role generation! Exit code of the 'oscap' process was %1.").arg(process.exitCode()).toUtf8().constData());
```

Comment:
Maybe it could be nice to copy the oscap's stderr to make the report more specific.

Reply:
I agree, but there is the code that does that already in the workbench.
Unfortunately, it is interwoven in the `OscapScannerLocal` instance and used in the `OscapScannerLocal::evaluate` method in a way that it is not easy to extract and use elsewhere.
I propose to leave it like it is for now and make the clean fix as another PR.

Reply:
@matejak OK, I agree, let's fix that later in another pull request. Thank you for investigation.

Reply:
I have squeezed a simple fix good-enough for the time being.